src/shape.cpp

```diff
@@ -490,6 +490,26 @@ bool shape::scalar() const
 
 bool shape::standard() const { return impl->m_standard; }
 
+
+bool shape::compatible_lens(const shape& s2) const
+{
+    if(dynamic() or s2.dynamic()) return true;
+    if(lens() != s2.lens() or type() != s2.type()) return false;
+    
+    // Lens must be the same; strides must be same except that
+    // axes with len=1 don't matter
+    for(size_t ind = 0; ind < lens().size(); ind++)
+    {
+        size_t l_ind(lens()[ind]);
+        if(l_ind != s2.lens()[ind] or
+          (l_ind != 1 and strides()[ind] != s2.strides()[ind]))
+            return false;
+    }
```

Comment:
No need to for this loop, if its both standard shape then just check that the lens are equal otherwise return false:

```cpp
static bool is_compatible_shape(const shape& actual, const shape& expected)
{
    // Check subshapes
    if(expected.type() == shape::tuple_type)
        return equal(actual.sub_shapes(), expected.sub_shapes(), &is_compatible_shape);
    // Only the expected can be dynamic
    if(expected.dynamic())
        return true;
    if(actual == expected)
        return true;
    if(actual.type() != expected.type())
        return false;
    if(actual.standard() and expected.standard())
        return actual.lens() == expected.lens();
    return false;
}
```

Reply:
It's still not clear to me what you mean by "actual" and "expected"  here.  I need to know what order to place the arguments when calling this, since only `expected` can be dynamic.

Reply:
Do we know that the recursion will always exit?  i.e. that `sub_shapes()` will always end with a shape that is not `tuple_type`?

Reply:
> It's still not clear to me what you mean by "actual" and "expected" here.

The "actual" is the shape being used at runtime, and the expected is the computed shape in the IR. For the case of dynamic shape the comptued IR will have a dynamic shape, but at runtime it will be a static shape, so we should just check that the type matches(and probably the ndim as well).